arro3-io/src/fsspec.rs

```diff
@@ -0,0 +1,125 @@
+use std::sync::Arc;
+
+use bytes::{Buf, Bytes};
+use futures::future::{BoxFuture, FutureExt};
+use parquet::arrow::async_reader::AsyncFileReader;
+use parquet::errors::ParquetError;
+use parquet::file::metadata::ParquetMetaDataReader;
+use parquet::file::FOOTER_SIZE;
+use pyo3::intern;
+use pyo3::prelude::*;
+use pyo3::types::{PyList, PyTuple};
+
+/// A wrapper around an Async fsspec filesystem instance
+pub struct AsyncFsspec {
+    fs: PyObject,
+    path: String,
+    file_length: usize,
+}
+
+impl AsyncFsspec {
+    pub fn new(fs: PyObject, path: String, file_length: usize) -> Self {
+        Self {
+            fs,
+            path,
+            file_length,
+        }
+    }
+}
+
+impl AsyncFileReader for AsyncFsspec {
+    fn get_bytes(
+        &mut self,
+        range: std::ops::Range<usize>,
+    ) -> BoxFuture<'_, parquet::errors::Result<bytes::Bytes>> {
+        async move {
+            let py_bytes_list = Python::with_gil(|py| -> PyResult<_> {
+                let paths = PyList::new_bound(py, vec![self.path.as_str()]);
+
+                let args = PyTuple::new_bound(
+                    py,
+                    vec![
+                        paths.to_object(py),
+                        range.start.into_py(py),
+                        range.end.into_py(py),
+                    ],
+                );
+
+                let coroutine =
+                    self.fs
+                        .call_method_bound(py, intern!(py, "_cat_ranges"), args, None)?;
+                pyo3_asyncio_0_21::tokio::into_future(coroutine.bind(py).clone())
+            })
+            .map_err(|err| ParquetError::External(Box::new(err)))?
+            .await
+            .map_err(|err| ParquetError::External(Box::new(err)))?;
+
+            let mut buffers = Python::with_gil(|py| py_bytes_list.extract::<Vec<Vec<u8>>>(py))
+                .map_err(|err| ParquetError::External(Box::new(err)))?;
```

Comment:
Extracting as `Vec<u8>` is unfortunate because it means we have a copy of the buffer fro Python to Rust, but I don't think this is avoidable.

Reply:
We are talking a few to tens of items, I am not worried by this.

Reply:
It'll be doing a full memory copy of whatever data was fetched, but still, I agree it's fine.

Reply:
Sorry, I was on the wrong thread and thought we were talking about rg/col selectors.
Here, you should be able to use python-backed bytes buffers again, no? The memory will not end up "owned" by rust, because it needs to be decompressed/decoded.

Reply:
> Here, you should be able to use python-backed bytes buffers again, no? The memory will not end up "owned" by rust, because it needs to be decompressed/decoded.

No, we can't because there are two different buffer objects. Here the traits require [`Bytes`](https://docs.rs/bytes/latest/bytes/struct.Bytes.html) objects, whereas arrow buffers are one step removed from that: https://github.com/apache/arrow-rs/blob/4adbeee14c40c080f8fca3bd42c0c3856bbb151c/arrow-buffer/src/bytes.rs#L28-L47

> This structs' API is inspired by the `bytes::Bytes`, but it is not limited to using rust's
> global allocator nor u8 alignment.

We can't construct a `bytes::Bytes` from external memory.

Reply:
pyarrow can (C++ pointers?), shame!

Reply:
I mean Rust can too, as evidenced by the `arrow::buffer::Buffer` object, but not every API takes that object. Most of the Rust I/O ecosystem works with `bytes::Bytes` because that's what `tokio` uses, and they generally don't need to work with externally-allocated memory.